supporting_scripts/generate_code_cov_table/generate_code_cov_table.py

```diff
@@ -31,29 +34,89 @@ def environ_or_required(key, required=True):
     )
 
 
-def get_client_tests_cov_report_url(build_id):
-    return f"{base_url}/browse/{project_key}-{plan_key}{build_id}/latest/artifact/shared/Coverage-Report-Client-Tests"
+def download_and_extract_zip(url, headers):
+    if url is None:
+        return None
+    try:
+        response = requests.get(url, headers=headers)
+        response.raise_for_status()
+
+        try:
+            zip_test = zipfile.ZipFile(BytesIO(response.content))
+            zip_test.close()
+
+            return BytesIO(response.content)
+        except zipfile.BadZipFile:
+            logging.error("The downloaded content is not a valid ZIP file.")
+            return None
+    except requests.RequestException as e:
+        logging.error(f"Failed to download the file: {e}")
+        return None
+
 
+def get_html_content(zip_file_bytes, file_name):
+    try:
+        with zipfile.ZipFile(zip_file_bytes, 'r') as zip_ref:
+            with zip_ref.open(file_name + '.html') as file_data:
+                return file_data.read().decode('utf-8')
+    except KeyError:
+        logging.error(f"File {file_name} not found in the archive.")
+        return None
+    except zipfile.BadZipFile:
+        logging.error("The provided file does not appear to be a valid zip file.")
+        sys.exit(1)
 
-def get_server_tests_cov_report_url(build_id):
-    return f"{base_url}/browse/{project_key}-{plan_key}{build_id}/latest/artifact/shared/Coverage-Report-Server-Tests"
 
+def get_artifacts_of_the_last_completed_run(headers, branch, run_id):
+    response = requests.get(runs_url, headers=headers, params={'branch': branch})
 
-def get_latest_build_id(username, password, branch_name):
-    bamboo_branch_name = branch_name.replace("origin/", "").replace("/", "-")
-    url = f"{base_url}/rest/api/latest/plan/{project_key}-{plan_key}/branch/{bamboo_branch_name}.json"
-    response = requests.get(url, auth=(username, password))
     if response.status_code == 200:
-        branch = json.loads(response.content)
-        return int(branch["shortKey"].replace(plan_key, ""))
+        runs = response.json()['workflow_runs']
+
+        completed_runs = list(run for run in runs if run['status'] == 'completed' and run['name'] == 'Test')
+        if len(completed_runs) == 0:
+            logging.error("No completed test runs found.")
+            sys.exit(1)
+
+        if run_id is None:
+            latest_completed_workflow = max(completed_runs, key=lambda x: x['created_at'], default=None)
+            artifacts_url = latest_completed_workflow['artifacts_url']
+        else:
+            filtered_runs = [run for run in runs if str(run['id']) == str(run_id)]
+            if len(filtered_runs) == 0:
+                logging.error("No run found with the specified ID.")
+                sys.exit(1)
+            elif len(filtered_runs) > 1:
+                logging.error("Multiple runs found with the same ID. ID should be unique.")
+                sys.exit(1)
```

Comment:
I don't quite get why we do this, if we somehow get multiple runs I would at most notify the user but generate the coverage table anyways with the first result.